src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -244,6 +247,50 @@ public final SocketAddress remoteAddress() {
         return (T) originalRequest;
     }
 
+    /**
+     * Resolves the specified {@code promise} with the specified {@code result}. If {@code promise} is done
+     * already, this method does the following:
+     * <ul>
+     *   <li>Log at WARN level about the failure, and</li>
+     *   <li>Release {@code result} if it is {@linkplain ReferenceCounted a reference-counted object},
+     *       such as {@link ByteBuf} and {@link FullHttpResponse}.</li>
+     * </ul>
+     */
+    public void resolvePromise(Promise<?> promise, Object result) {
+        @SuppressWarnings("unchecked")
+        final Promise<Object> castPromise = (Promise<Object>) promise;
+
+        if (castPromise.trySuccess(result)) {
+            // Resolved successfully.
+            return;
+        }
+
+        try {
+            if (!(promise.cause() instanceof TimeoutException)) {
```

Comment:
Or maybe they're already logged in timeout handler? If so then it's ok, I guess it's very exceptional to even get to this code.